python/rapidsmpf/rapidsmpf/rmm_resource_adaptor.pyx

```diff
@@ -0,0 +1,206 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+
+from cython.operator cimport dereference as deref
+from libc.stdint cimport uint64_t
+from libcpp.cast cimport dynamic_cast
+from rmm.pylibrmm.memory_resource cimport (DeviceMemoryResource,
+                                           UpstreamResourceAdaptor)
+
+# Pointer alias used by `dynamic_cast`, which doesn't accept pointer types.
+ctypedef cpp_RmmResourceAdaptor* cpp_RmmResourceAdaptor_ptr
+
+cdef class ScopedMemoryRecord:
+    cdef cpp_ScopedMemoryRecord _handle
+
+    def __cinit__(self):
+        self._handle = cpp_ScopedMemoryRecord()
```

Comment:
For safety, I think we should also add `__dealloc__` that just creates a new object from the default constructor, like [here](https://github.com/rapidsai/rapidsmpf/blob/b5b669f20762d9368b3a1310af1d5969a91a1e38/python/rapidsmpf/rapidsmpf/config.pyx#L37-L39).

Reply:
I guess, we would also need to do:
```python
    def __cinit__(self):
        with nogil:
            self._handle = cpp_ScopedMemoryRecord()
```

But `ScopedMemoryRecord` is just a collection of counters